### ISSUE TITLE
IntelliJ "Annotate" action sometimes fails with truncated output

### ISSUE THREAD

**Issue by tMaxx** [andy-5/wslgit#41]
#### Background

While working on a Git project I noticed that when trying to annotate a file (GUI equivalent of `git blame`), one of three things can happen, depending on file being annotated:

1. File gets annotated without any problems,
2. An error is shown with message "Number of lines annotated by Git is not equal to number of lines in the file. Check file encoding and line separators"; when clicking to display it anyway it is truncated at around half the length of said file,
3. IDE errors out with message along the lines of "Error: Unexpected text end at 24566".

Java unit test files seem to cause the most trouble, probably due to their file/line length, and number of contributors/revisions.

Probably similar to #28...?

#### System info

wslgit version 0.6.0 official .exe, not additionally configured in any way

WSL: Ubuntu 18.04, Windows 10 Pro build 1803, 17134.228  
Default shell changed to zsh

-------

I haven't tried any debugging (to see which command is actually issued to wslgit), or setting `WSLGIT_USE_INTERACTIVE_SHELL=false` yet (don't know if it will even make sense in this case).

**Comment by andy-5** [comment]
Thanks for your report.

First, it's nice to see that wslgit seems to generally work with IntelliJ, because there is an older issue (#9) which suggests that it does not work with related tools like PHPStorm and Android Studio. Did it require any special configuration in addition to configuring the git executable?

Regarding your issue, I can only guess. Do the files showing the wrong behavior maybe have mixed line endings (i.e. `\n` and `\r\n`). This could explain the mismatch. There is also a [help page](https://www.jetbrains.com/help/idea/2017.1/handling-lf-and-crlf-line-endings.html) for dealing with different line endings and Git in IntelliJ.

The best way to debug this is probably to find out the exact command that is executed by IntelliJ and then run this command manually in a terminal. Then you can compare the output of Git to the lines shown in IntelliJ.

**Comment by tMaxx** [comment]
> Did it require any special configuration in addition to configuring the git executable?

No additional configuration was required.

Sure, for some functions it sometimes fails (noticed today additional "Failed to generate diff" error), but what is important (commits, merges, branching, push/pull, setting identity for commit) works without any problem whatsoever.

> Do the files showing the wrong behavior maybe have mixed line endings

No, LF all the way throughout all projects that I'm currently working on.

I might have some free time to investigate further in the next week or two, but since it's not a complete dealbreaker for me I won't be giving it the highest priority.

**Comment by andy-5** [comment]
@mbeccati So this issue is also present if you call git inside WSL directly from a .bat file? That suggests that the problem is likely elsewhere, not in wslgit. But it would still be interesting to know what is going on.

**Comment by andy-5** [comment]
@tMaxx That sounds good, thanks for investigating.
Yes, it would be great to have the exact commands (with inputs) to replicate the issue outside of IntelliJ and compare it with other environments.

Maybe the specific byte sequences you mentioned are messed up somewhere in the conversion/translation between WSL and Windows.

**Comment by mbeccati** [comment]
@andy-5 yes, precisely. If I redirect the git output to a file it doesn't seem to be truncated though. This thing remains a mystery to me... I've had to install git for windows to work around the issue.

**Comment by alekseymvt** [comment]
> First, it's nice to see that wslgit seems to generally work with IntelliJ, because there is an older issue (#9) which suggests that it does not work with related tools like PHPStorm and Android Studio.?

Offtop: WslGit working with PHPStorm 2018.2.4.
Windows 1809. Ubuntu 18.04.

**Comment by tMaxx** [comment]
It seems to be working fine seemingly for all cases that weren't working before - annotating `wslgit.rs` doesn't trigger an error anymore. And I can't remember when I last saw an error like that on my work laptop.

Windows 10 Pro 17134.407
IntelliJ IDEA 2018.3.1

Haven't updated wslgit since first installing it, so it may not have been its fault to begin with.

**Comment by mbeccati** [comment]
@tMaxx I thought we got rid of it, but I got the issue again just a few minutes ago. Suddenly I saw 3/4th of the open file with the green vertical left bar as if I had just added all those lines. Annotating brought up the error again...

**Comment by bacalov** [comment]
Restarting the IDE worked for me.
Rubymine 2018.3.5

**Comment by BimuratMukhtar** [comment]
similar issue from AndroidStudio 4.0 

local version: 
![image](https://user-images.githubusercontent.com/32272243/83394418-5deadf00-a41a-11ea-9e61-736701f82d7c.png)

gitlab:
![image](https://user-images.githubusercontent.com/32272243/83394509-8b378d00-a41a-11ea-8bd8-ed1bf9fdc7c0.png)

**Comment by BimuratMukhtar** [comment]
In my case on Android Studio 4.0 versions gitlab vs local differs, when trying to annotate similar issue
